### ISSUE TITLE
Fastest Path with Secondary Constraint

### ISSUE THREAD

**Issue by mattrjackson** [valhalla/valhalla#2277]
Hi,

I currently have a  use case that involves routing vehicles around a time decaying hazard.  In order to do this, what I need to calculate is the fastest route from A to B without exceeding a secondary constraint - I don't want to avoid the entire hazard in most cases, only avoid exceeding a given risk level when traversing the route.  Is there a recommended way to accomplish this with Valhalla? From my initial examination of documentation/code, it appears quite straightforward to minimize the total cost, but I didn't see an option for applying a secondary constraint. 

Thanks,

Matt

**Comment by kevinkreiser** [comment]
hi @mattrjackson i have a few questions to make sure i am understanding your question.

can you be more specific in what you mean by `time decaying hazard`. do you mean you have a road closure that is active from 11am to 1pm for example?

it seems to me that your hazard is not binary (fully avoided). is it something like radiation exposure that you are trying to not exceed as a secondary constraint?

is the secondary constraint cumulative? that is, does it accumulate as the route expansion progresses? at the moment the only things we accumulate are time and cost, and none of these are capped by a maximum value we just try to minimize them. in otherwords the expansion of a particular path doesnt stop when they reach a certain value.

one could add this functionality to the costing object. it has an allowed method that could return false if a maximum was reached. one would also have to extend the costing object to track this other constraint and set it in the cost object so that later calls could decide to disallow certain expansion. then there is the question of whether this measurement shoudl be passive or active. ie should we just measure it and stop expansion or should we also apply it to the cost to make sure it stays below a certain level.

i think a good example of this would be fuel usage. one could imagine the user specifies how much fuel they would want to spend on a trip and we need to both measure how much we estimate is used but also modify the cost to favor paths that will actually complete the route with the given amount of fuel.

i hope some of this information helps. long story short we dont support it at the moment but i could envision adding support for such a thing.

**Comment by mattrjackson** [comment]
Kevin,

Thanks for your response.  Yes, a radiation hazard is one application, but it's generally applicable to hazards from industrial spills, minimizing smoke inhalation traveling downwind of large fires, etc.  Fuel economy is also a potential use case.  It isn't merely a road/edge that is only available during certain periods of time. 

The secondary constraint is cumulative (it would be accrued dose in the case of a radiation hazard); on the forward search side it's more or less an exact calculation of the dose. On the reverse search side you have to approximate the dose, because you don't know apriori when you will arrive at your destination.  It's more of a passive constraint, as the ultimate objective remains to minimize distance or time, while not exceeding the constraint. 

I considered handling it in the costing object as you suggested. I could track within the costing class the total accrued dose at each edge, then use the allowed function to reject further expansion if the cumulative dose in the preceding edge exceeded the threshold. One question is how to handle the final step when the forward/backward searches meet? 

An alternative approach (and perhaps more robust in the long term?) might be to create an abstract base cost class with virtual members for retrieving cost, setting cost, adding, subtracting, defining the heuristic, and comparing cost structures (and maybe one or two other things I'm not thinking of). The current Cost structure could then inherit from the BaseCost class, but it would provide a convenient means to track the data needed to feed more advanced cost functions. I did a little digging into this approach yesterday, and it appears doable, but there are a few places where distances are added to cost in some of the thor algorithms that would need to be addressed in some form.

It's something that if it sounds like something that might be of interest to the project, I do have some time to work on it, and would gladly commit it back for you to use as you deem fit. 

Matt

**Comment by kevinkreiser** [comment]
so we have Costing classes and a Cost class. as far as Costing goes we already have an abstract base cost class called `DynamicCost` that all the costings implement. i do believe this is the right place to add the heuristics for how much "secondary cost" a particular edge or node contributes. However you are correct, you need to accumulate and store this "secondary cost" in the Cost class. if it suites your needs to extend that class i would encourage you to do so and open a draft PR. as you make progress i would be glad to answer questions and give suggestions about how to achieve what you are looking for! let me know if you'd like to do anymore more planning about the implementation before you start, ill be glad to answer any more questions here!

**Comment by mattrjackson** [comment]
As a quick test, I just added a vector<double> to the cost class to store any variables that will be needed to evaluate secondary constraints, and updated the addition/subtraction/multiplication overrides accordingly. If I am looking at things correctly, it does appear that in order to fully enforce the secondary constraints correctly, one will need to update the SetForwardConnection/SetReverseConnection functions in the bidirectional A* algorithm to analyze the cost struct, and verify that no secondary constraints are exceeded when joining the two paths - otherwise just using the existing allowed function, as you suggest, is sufficient to ensure the constraint(s) are not exceeded on one side of the search or the other. It's somewhat uncommon for the secondary constraint to be significantly impacted by both sides of the search, but not impossible. 

I'm thinking about adding an EvaluateConstraints method to dynamiccost to pass the total cost and evaluate whether it is acceptable or not, this could be a default call from the Allowed function, but then could also be queried by thor to determine whether a series of secondary constraints are satisfied when evaluating the total cost of the path. Ultimately it will require a (hopefully small) number of changes in thor to accommodate - I haven't fully explored the extent yet. Does that make sense? If there's additional places (beyond the bidirectional a* class) that should be made for consistency sake, I do need to understand that as well.

If yes, then I'll wrap up the changes (hopefully in the next week or two), and construct some unit tests - should be relatively easy to create a simple radial function that decays as a function of time to verify everything works as expected. Will plan to open a PR at that point for you/others to review.

**Comment by kevinkreiser** [comment]
@mattrjackson yep, you will need to update the forward/reverse connection setting in the algorithm. with respect to the changes to the base costing class to do the total cost thresholding it sounds like its something that could work. in addition to that there are a couple of things i would suggest:

1. dont worry about the other algorithms just yet, if bidirectional a* suites your needs no need to expand your work until you've proven what you wanted to do there. we can do the "consistency" work as a second pass once we've fully vetted the approach
2. since this stuff is called in a tight loop we will need to check performance considerations. we can help you with this once its ready.
3. we will need some design on the interface to this functionality, i think its fine if you want to hardcode some values to start with but at some point we'll need to wire this through the request parsing etc.
4. when it comes to unit tests it could be beneficial for you to try out our new `gurka` functionality, which allows you to make a fake road network and run routes on it. for more information, have a look at the readme: https://github.com/valhalla/valhalla/blob/master/test/gurka/README.md

also, its great to hear that you've made a bunch of progress already! let me know as things come up if you have more questions and thanks for contributing :+1:

**Comment by mattrjackson** [comment]
1. Sounds good. I'll focus on the bidirectional A* for now.
2. Yes, I certainly don't want to burden things - vector is probably overkill for this task - realistically I doubt anyone would want to evaluate more than 2-3 constraints, so a fixed array might be a better solution.
3.  I've mostly done this already - I created a CostingConstraint message to hold the data, and have tested it through the run_route program, and have verified it's taking a json string and parsing it correctly.
4. Will do. Seems like a good way to do some of the tests.

From a design perspective, there is one thing I'm concerned about right now:

I ended up having to add an EdgeCostReverse method to dynamiccost - the reason for this is that I need to handle differently the reverse search constraint calculation from the forward search, in terms of approximating the hazard. I have defaulted the implementation of EdgeCostReverse to call EdgeCost, so it shouldn't break any existing implementations, if another user has made their own custom costing functions.

That said, I would like to implement EdgeCostReverse for the existing costing functions (basically a copy-paste of the EdgeCost, with the addition of a conditional block that checks for the existence of constraints, and if so, calling the appropriate method). By doing this, it gives the developer the flexibility to apply secondary constraint(s) to any of the existing costing functions by doing nothing more than setting the function pointers and passing the constraints in as input. This seems preferable to creating a series of new costing functions that behave identically to the current ones, with the addition of 3-4 lines of code in the EdgeCost/TransitionCost algorithms.

If there's a way to determine whether a given directededge is from a forward or reverse search, it would eliminate the need for the new method entirely, but absent that option, this was the least disruptive way to add the functionality that I could think of. 

I did consider your idea of using the Allowed method to handle the secondary constraints, but the problem is I do need to store the values of the constraints on the cost. There are some hacks that could be used to track forward/backward search by setting a flag in conjunction with the Allowed routine, but I wanted to avoid those if possible, as it then sets a dependency on them being called in a specific order, which could result in some nasty bugs down the road.
